File: Python/First_Unique_Char.py
# Question: Given a string, find the first character that is unique
# Solution: Count the occurance of each letter, find the first letter with 1 occurance from your hashmap
#           because we don't want to loop over our input string twice if it's very long
# Difficulty: Easy

def firstUniqChar(s: str) -> int:
        # Create a hashmap to store the counts of each letter as well as the index
        letterCounts = {}

        # Store the lowest index we see with a letter count of 1 when we traverse our hashmap
        smallestIndex = len(s)

        # Store each items count as well as the first index it occured at
        for i, v in enumerate(s):
            getChar = letterCounts.get(v, [0, i])
            letterCounts[v] = (getChar[0] + 1, getChar[1])

        # Loop through the hashmap updating the smallestIndex each time we see a letter
        # with a count of 1 and a smaller index than the previous 
        for i, v in enumerate(letterCounts):
            if letterCounts[v][0] == 1:
                smallestIndex = min(smallestIndex, letterCounts[v][1])
                        
        # Return either the index of the first char or -1 if there is no unique char
        return smallestIndex if smallestIndex != len(s) else -1

File: Go/First_Unique_Char.go
import "math"

func firstUniqChar(s string) int {
	// Create a struct which we can populate our hashmap with
	type Pair struct {
		count int
		index int
	}

	seen := map[rune]*Pair{}

	// Loop over the string, incrementing count if the char exists and adding it othwerwise
	for i, v := range s {
		if _, ok := seen[v]; ok {
			seen[v].count++
		} else {
			seen[v] = &Pair{1, i}
		}
	}

	lowest := math.Inf(1)

	// Loop over the hashmap, and find the minimum index with a count of 1
	for _, v := range seen {
		if v.count == 1 {
			lowest = math.Min(lowest, float64(v.index))
		}
	}

	if lowest == math.Inf(1) {
		return -1
	} else {
		return int(lowest)
	}
}